File: api/src/new_phone/schemas/billing.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field

# ── Usage Records ─────────────────────────────────────────


class UsageRecordCreate(BaseModel):
    period_start: datetime
    period_end: datetime
    metric: str = Field(min_length=1, max_length=50)
    quantity: float = Field(ge=0)
    unit_cost: float | None = None
    total_cost: float | None = None


class UsageRecordResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    tenant_id: uuid.UUID
    period_start: datetime
    period_end: datetime
    metric: str
    quantity: float
    unit_cost: float | None
    total_cost: float | None
    created_at: datetime
    updated_at: datetime


# ── Rate Decks ────────────────────────────────────────────


class RateDeckCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    description: str | None = None
    is_default: bool = False
    is_active: bool = True


class RateDeckUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=100)
    description: str | None = None
    is_default: bool | None = None
    is_active: bool | None = None


class RateDeckResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    tenant_id: uuid.UUID
    name: str
    description: str | None
    is_default: bool
    is_active: bool
    created_at: datetime
    updated_at: datetime


# ── Rate Deck Entries ─────────────────────────────────────


class RateDeckEntryCreate(BaseModel):
    prefix: str = Field(min_length=1, max_length=20)
    destination: str = Field(min_length=1, max_length=100)
    per_minute_rate: float = Field(ge=0)
    connection_fee: float = Field(ge=0, default=0.0)
    minimum_seconds: int = Field(ge=0, default=0)


class RateDeckEntryResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    rate_deck_id: uuid.UUID
    prefix: str
    destination: str
    per_minute_rate: float
    connection_fee: float
    minimum_seconds: int
    created_at: datetime
    updated_at: datetime


# ── Rate Lookup Result ────────────────────────────────────


class RateLookupResponse(BaseModel):
    dialed_number: str
    matched_prefix: str | None
    destination: str | None
    per_minute_rate: float | None
    connection_fee: float | None
    minimum_seconds: int | None


# ── Billing Config ────────────────────────────────────────


class BillingConfigUpdate(BaseModel):
    billing_provider: str | None = Field(None, pattern=r"^(connectwise|pax8|manual)$")
    connectwise_agreement_id: str | None = None
    pax8_subscription_id: str | None = None
    billing_cycle_day: int | None = Field(None, ge=1, le=28)
    auto_generate: bool | None = None


class BillingConfigResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    tenant_id: uuid.UUID
    billing_provider: str
    connectwise_agreement_id: str | None
    pax8_subscription_id: str | None
    billing_cycle_day: int
    auto_generate: bool
    created_at: datetime
    updated_at: datetime

File: web/src/api/billing.ts
import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query"
import { api } from "./client"
import { useAuthStore } from "@/stores/auth-store"
import { queryKeys } from "./query-keys"

export interface UsageRecord {
  id: string
  tenant_id: string
  period_start: string
  period_end: string
  metric: string
  quantity: number
  unit_cost: number | null
  total_cost: number | null
  created_at: string
  updated_at: string
}

export interface RateDeck {
  id: string
  tenant_id: string
  name: string
  description: string | null
  is_default: boolean
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface RateDeckEntry {
  id: string
  rate_deck_id: string
  prefix: string
  destination: string
  per_minute_rate: number
  connection_fee: number
  minimum_seconds: number
  created_at: string
  updated_at: string
}

export interface BillingConfig {
  id: string
  tenant_id: string
  billing_provider: string
  connectwise_agreement_id: string | null
  pax8_subscription_id: string | null
  billing_cycle_day: number
  auto_generate: boolean
  created_at: string
  updated_at: string
}

export interface RateDeckCreate {
  name: string
  description?: string
  is_default?: boolean
  is_active?: boolean
}

export interface RateDeckEntryCreate {
  prefix: string
  destination: string
  per_minute_rate: number
  connection_fee?: number
  minimum_seconds?: number
}

export function useUsageRecords(periodStart?: string, periodEnd?: string) {
  const tenantId = useAuthStore((s) => s.activeTenantId)
  const params = new URLSearchParams()
  if (periodStart) params.set("period_start", periodStart)
  if (periodEnd) params.set("period_end", periodEnd)
  return useQuery({
    queryKey: queryKeys.billing.usage(tenantId!, periodStart, periodEnd),
    queryFn: () => api.get(`/tenants/${tenantId}/billing/usage?${params}`).then((r) => r.data as { items: UsageRecord[]; total: number }),
    enabled: !!tenantId,
  })
}

export function useRateDecks() {
  const tenantId = useAuthStore((s) => s.activeTenantId)
  return useQuery({
    queryKey: queryKeys.billing.rateDecks(tenantId!),
    queryFn: () => api.get(`/tenants/${tenantId}/billing/rate-decks`).then((r) => r.data as RateDeck[]),
    enabled: !!tenantId,
  })
}

export function useCreateRateDeck() {
  const tenantId = useAuthStore((s) => s.activeTenantId)
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (data: RateDeckCreate) => api.post(`/tenants/${tenantId}/billing/rate-decks`, data).then((r) => r.data as RateDeck),
    onSuccess: () => qc.invalidateQueries({ queryKey: queryKeys.billing.rateDecks(tenantId!) }),
  })
}

export function useBillingConfig() {
  const tenantId = useAuthStore((s) => s.activeTenantId)
  return useQuery({
    queryKey: queryKeys.billing.config(tenantId!),
    queryFn: () => api.get(`/tenants/${tenantId}/billing/config`).then((r) => r.data as BillingConfig),
    enabled: !!tenantId,
  })
}

export function useUpdateBillingConfig() {
  const tenantId = useAuthStore((s) => s.activeTenantId)
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (data: Partial<BillingConfig>) => api.put(`/tenants/${tenantId}/billing/config`, data).then((r) => r.data as BillingConfig),
    onSuccess: () => qc.invalidateQueries({ queryKey: queryKeys.billing.config(tenantId!) }),
  })
}
